### ISSUE TITLE
Add support for OIDC `prompt` key

### ISSUE THREAD

**Issue by farcaller** [kanidm/kanidm#2988]
### Is your feature request related to a problem? Please describe

I want to be able to force users to log in again when they access a specific oauth app.

### Describe the solution you'd like

[OIDC](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) defines a prompt parameter, that communicates the request to relogin via `prompt=login`.

### Describe alternatives you've considered

None considered, we have one in the spec.

**Comment by Firstyear** [comment]
I like the idea of this, but the quirk here is in the OIDC specification when you set "prompt=login" and the AS doesn't understand that, then:

```
     If an OP receives a prompt value outside the set defined above that it does not understand, it MAY return an error or it MAY ignore it; in practice, not returning errors for not-understood values will help facilitate phasing in extensions using new prompt values. 
```

What this means is that as the OIDC client, when you issue a prompt=login, when you get the OK response you can not distinguish between "did they ignore the prompt or was the login actually correctly performed". 

So while we can pinky-promise to do the right thing, I am curious about what you want to achieve here with this feature suggestion.

**Comment by Firstyear** [comment]
(I'm not against adding it btw, I kind of like the idea, but I think OIDC half-arsed this part of the spec).

**Comment by yaleman** [comment]
I guess the assumption is that if it doesn't just straight up barf at the `prompt=login` the spec says to only pass them back if they auth, so it's ... the closest you can get to a pinky-swear, from the IdP that you ostensibly trust by configuring it into the applicaition

> The Authorization Server SHOULD prompt the End-User for reauthentication. If it cannot reauthenticate the End-User, **it MUST return an error, typically login_required.**

**Comment by Firstyear** [comment]
The assumption there is they *do* support it. If they don't they're free to return the 200 OK and you're none the wiser. So you have to actually trust the IDP did the thing, because there are no signals otherwise.

**Comment by Firstyear** [comment]
Also it's "should" not "must". So you know. lmao.

**Comment by yaleman** [comment]
Yeah, but if you can't trust your IdP then you're already in a bad spot 😄

**Comment by Firstyear** [comment]
True. Just me nitpicking and being a perfectionist :)

**Comment by farcaller** [comment]
> I am curious about what you want to achieve here with this feature suggestion.

As was correctly noted above, I want kanidm to deny login if the user failed to reauthenticate. Right now it (obviously) does the next best thing, which is ignoring the prompt altogether.

What I **actually** want to achieve is even shorter term login credentials to guarantee that the user tapped the hardware token for auth. I'd be happy with either the prompt feature, or having an option to enforce the same in the oauth client configuration (per client).

**Comment by Firstyear** [comment]
Yeah I think that sounds like a good feature. I don't have time now, but I can see it being useful soon.

**Comment by 9SMTM6** [comment]
My understanding from [this](https://auth0.com/docs/authenticate/login/max-age-reauthentication) is that prompt=login is more of an UX- feature.

One that I'd love to have, I came here after searching for why prompt=login didn't do anything, but not something you should use to ensure the user re-authenticated.

**Comment by Firstyear** [comment]
Like I said, it's a weak signal. We should add it but we don't have a great way to actually force the re-login today. The other issue is we have other current priorities so this is on the backburner.

**Comment by vnghia** [comment]
Hello I have a usecase for the `prompt` key that I think could be interesting. I am selfhosting a meeting software and when you open a meeting link `https://example/abc-defg-hij`, it will issue a `prompt=none` which should:
- Automatically log user in and then the user can join the meeting with their account.
- Or return an error code and then the user will join the meeting as a guest user (if that meeting is open for guest access of course).

I wonder could we support for that `prompt=none` as well ?

**Comment by Firstyear** [comment]
prompt=none is a lot easier to support in the current design.

**Comment by Firstyear** [comment]
During prompt=none I think the other thing is if the user isn't authenticated is it up to Kanidm to ask if they want to (and allows the escape hatch to proceed without authentication if the user is a guest), or is it up to the calling application to then have a separate "login" button that triggers an auth flow where prompt=none  is absent?

**Comment by vnghia** [comment]
I think it is up to the calling application. Kanidm should make it as automatic as possible, no manual intervention is needed. I've opened an issue here previously to check that with the meeting software: https://github.com/suitenumerique/meet/issues/708

**Comment by yaleman** [comment]
[OIDC Core 3.1.2.1](https://openid.net/specs/openid-connect-core-1_0.html?utm_source=chatgpt.com#AuthRequest)'s pretty clear, prompt=none is it's up to the app to deal with, the IdP just says "nope, go away" if you're not already authenticated.


> `prompt<snip>none`
> 
> **The Authorization Server MUST NOT display any authentication or consent user interface pages.** An error is returned if an End-User is not already authenticated or the Client does not have pre-configured consent for the requested Claims or does not fulfill other conditions for processing the request. The error code will typically be login_required, interaction_required, or another code defined in [Section 3.1.2.6](https://openid.net/specs/openid-connect-core-1_0.html?utm_source=chatgpt.com#AuthError). This can be used as a method to check for existing authentication and/or consent.


... and only redirect if it's valid otherwise throwing an error is fine. (because of the whole "bad redirect" issue.

**Comment by Firstyear** [comment]
Perfect, that makes it really easy then. @CEbbinghaus wanted to have a look so I think for prompt=none, we can assign it to them. For prompt=login, that's a separate issue still because it needsa  bit more work to achieve.
